Bugzilla/Install/AssetManager.pm

```diff
@@ -0,0 +1,224 @@
+    my $data = pack "H*", $hex;
+    return "sha$SHA_VERSION-" . encode_base64($data, "");
+}
+
+sub compile_file {
+    my ($self, $file) = @_;
+    return unless -f $file;
+    my $base_dir  = $self->base_dir;
+    my $asset_dir = $self->asset_dir;
+    my $asset_map = $self->state->{asset_map};
+
+    my $key = File::Spec->abs2rel( $file, $base_dir );
+    return if $asset_map->{$key};
+
+    if ($file =~ /\.(jpe?g|png|gif|ico|woff|js)$/i) {
+        my $ext            = $1;
+        my $digest         = $self->_digest_file($file);
+        my $asset_file     = File::Spec->catfile($asset_dir, "$digest.$ext");
+        cp($file, $asset_file);
+        if ($digest eq $self->_digest_file($asset_file)) {
+            $asset_map->{$key} = File::Spec->abs2rel($asset_file, $asset_dir);
+        }
+        else {
+            die "failed to write $asset_file";
+        }
+    }
+    elsif ($file =~ /\.css$/) {
+        my $content = read_file($file);
+
+        # minify
+        $content =~ s{(?<!=)url\(([^\)]+)\)}{$self->_css_url_rewrite($1, $file)}eig;
+        my $digest = $self->_digest_string($content);
+        my $asset_file     = File::Spec->catfile($asset_dir, "$digest.css");
+        write_file($asset_file, $content);
+        if ($digest eq $self->_digest_file($asset_file)) {
+            $asset_map->{$key} = File::Spec->abs2rel($asset_file, $asset_dir);
+        }
+        else {
+            die "failed to write $asset_file";
+        }
+    }
+}
+
+sub compile_all {
+    my ($self) = @_;
+    my $asset_map = $self->state->{asset_map};
+
+    %$asset_map = ();
+
+    my $wanted = sub {
+        $self->compile_file($File::Find::name);
+    };
+
+    find( { wanted => $wanted, no_chdir => 1 }, @{ $self->source_dirs });
+
+    $self->_save_state();
+}
+
+sub _css_url_rewrite {
+    my ($self, $url, $file) = @_;
+    my $dir = dirname($file);
+    # rewrite relative urls as the unified stylesheet lives in a different
+    # directory from the source
+    $url =~ s/(^['"]|['"]$)//g;
+    if ($url =~ m!^(/|data:)!) {
+        return 'url(' . $url . ')';
+    }
+    else {
+        my $url_file = File::Spec->rel2abs($url, $dir);
+        my $ref_file = File::Spec->abs2rel( $url_file, $self->base_dir );
+        $self->compile_file($url_file);
+        return sprintf( "url(%s)", $self->asset_file($ref_file, $self->asset_dir));
+    }
+}
+
+sub _new_digest { Digest::SHA->new($SHA_VERSION) }
+
+sub _digest_file {
+    my ($self, $file) = @_;
+    my $digest = $self->_new_digest;
+    $digest->addfile($file);
+    return $digest->hexdigest;
+}
+
+sub _digest_string {
+    my ($self, $string) = @_;
+    my $digest = $self->_new_digest;
+    $digest->add($string);
+    return $digest->hexdigest;
+}
+
+sub _build_base_dir  { Cwd::realpath(File::Spec->rel2abs(bz_locations->{cgi_path})) }
+sub _build_asset_dir {
+    my ($self) = @_;
+    my $dir = Cwd::realpath(File::Spec->rel2abs(bz_locations->{assetsdir}));
+
+    if ($dir && -d $dir) {
```

Comment:
when is $dir undefined, and why would it be correct to return that as the asset_dir?